include/nbl/builtin/hlsl/shapes/beziers.hlsl

```diff
@@ -13,116 +16,226 @@ namespace hlsl
+            float_t calcArcLenNumeric(float_t t)
+            {
+                
+                double num = 0.0;
+                const int steps = 100;
+                    // from 0 to t
+                const double rcp = double(t) / double(steps);
+                for (int i = 0; i < steps; i++)
+                {
+                    double x = double(i) * rcp + 0.5 * rcp;
+                    num += sqrt(x * (lenA2 * 4.0 * x + AdotB * 4.0) + c) * rcp;
+                }
+                
+                return num;
+                
+            }
+            
+            float_t calcArcLenInverse(float_t arcLen, float_t accuracyThreshold, float_t hint, Quadratic<float_t> quadratic)
+            {
+                float_t xn = hint;
+
+                if (arcLen <= accuracyThreshold)
+                    return arcLen;
+
+                    // TODO: implement halley method
+                const uint32_t iterationThreshold = 32;
+                for(uint32_t n = 0; n < iterationThreshold; n++)
+                {
+                    float_t arcLenDiffAtParamGuess = arcLen - calcArcLen(xn);
 
-        static QuadraticBezierOutline construct(float2 a, float2 b, float2 c, float thickness)
+                    if (abs(arcLenDiffAtParamGuess) < accuracyThreshold)
+                        return xn;
+
+                    float_t differentialAtGuess = length(2.0*quadratic.A * xn + quadratic.B);
+                        // x_n+1 = x_n - f(x_n)/f'(x_n)
+                    xn -= (calcArcLen(xn) - arcLen) / differentialAtGuess;
+                }
+
+                return xn;
+            }
+
+        };
+        
+        static Quadratic construct(float2_t A, float2_t B, float2_t C)
+        {            
+            Quadratic ret = { A, B, C };
+            return ret;
+        }
+        
+        static Quadratic constructFromBezier(QuadraticBezier<float_t> curve)
         {
-            QuadraticBezier bezier = { a, b, c };
-            QuadraticBezierOutline ret = { bezier, thickness };
+            const float2_t A = curve.P0 - 2.0 * curve.P1 + curve.P2;
+            const float2_t B = 2.0f * (curve.P1 - curve.P0);
+            const float2_t C = curve.P0;
+            
+            Quadratic ret = { A, B, C };
             return ret;
         }
-
-        // original from https://www.shadertoy.com/view/lsyfWc
-        float ud(float2 pos)
+        
+        float2_t evaluate(float_t t)
         {
-            const float2 A = bezier.A;
-            const float2 B = bezier.B;
-            const float2 C = bezier.C;
-                    
+            return t * (A * t + B) + C;
+        }
+        
+        struct DefaultClipper
+        {
+            static DefaultClipper construct()
+            {
+                DefaultClipper ret;
+                return ret;
+            }
+        
+            inline float2_t operator()(const float_t t)
+            {
+                return clamp(t, 0.0, 1.0);

```

Comment:
does this return `(clamp(t, 0.0, 1.0), clamp(t, 0.0, 1.0))` or `(clamp(t, 0.0, 1.0), 0)`?

Reply:
idk, will change it to `clamp(t, 0.0, 1.0).xx`